ruby/lib/minitest/queue/test_time_reporter.rb

```diff
@@ -0,0 +1,69 @@
+require 'minitest/reporters'
+
+module Minitest
+  module Queue
+    class TestTimeReporter < Minitest::Reporters::BaseReporter
+      include ::CI::Queue::OutputHelpers
+
+      def initialize(build:, limit: nil, percentile: nil, **options)
+        super(options)
+        @test_time_hash = build.fetch
+        @limit = limit
+        @percentile = percentile
+        @success = true
+      end
+
+      def report
+        return if limit.nil? || test_time_hash.empty?
+
+        puts '+++ Test Time Report'
+
+        if offending_tests.empty?
+          msg = "The #{humanized_percentile} of test execution time is within #{limit} milliseconds."
+          puts green(msg)
+          return
+        end
+
+        @success = false
+        puts <<~EOS
+          #{red("Detected #{offending_tests.length} test(s) over the desired time limit.")}
+          Please make them faster than #{limit}ms in the #{humanized_percentile} percentile.
+        EOS
+        offending_tests.each do |test_name, duration|
+          puts "#{red(test_name)}: #{duration}ms"
+        end
+      end
+
+      def success?
+        @success
+      end
+
+      def record(*)
+        raise NotImplementedError
+      end
+
+      private
+
+      attr_reader :test_time_hash, :limit, :percentile
+
+      def humanized_percentile
+        percentile_in_percentage = percentile * 100
+        "#{percentile_in_percentage.to_i}th"
+      end
+
+      def offending_tests
+        @offending_tests ||= begin
+          test_time_hash.each_with_object({}) do |(test_name, durations), offenders|
+            duration = calculate_percentile(durations)
+            next if duration <= limit
+            offenders[test_name] = duration
+          end
+        end
+      end
+
+      def calculate_percentile(array)
+        array.sort[(percentile * array.length).ceil - 1]
```

Comment:
Do we know beforehand how many items exist in the result array? Calculating a percentile value only makes sense when the sample set is large enough, e.g. anything under 3 items is usually not meaningful.

Reply:
It depends on how someone uses the grind command. You can configure how often you want the test to be grinded.

Reply:
In core, we run each test 1000 times. This should be enough for percentile.

Reply:
Right. For general use though, I think it would make sense to print a warning if less than 3 items exist in the array. It's generally considered wrong to calculate percentile scores for 1 or 2 items in an array.

Reply:
I think that makes sense.

Reply:
Since the command is used for grind, it is assumed that the number is big enough that the grind will give you an accurate result. What I am trying to say is that if there is a minimum number for percentile, we should have a minimum number for grind.

Reply:
But the present of a minimum number of grind will warrant an accurate percentile.

Reply:
That's why I was asking "Do we know beforehand how many items exist in the result array?". If we know that with certainty (i.e., the check is already somewhere else), I'm fine with that too.

Reply:
It might be ok to discard a grind if it is less than 50 runs. That is probably the easiest way to do that.

Reply:
> Since the command is used for grind

If the code allows for other use cases (not tied to grind), it should independently do those checks and output some warnings.

Reply:
In case in the future we face redis issues to store this data you can refactor to use a bucket approach: https://stackoverflow.com/a/1249003/1542719

Anyway, we should monitor the redis instance as this will probably increase the redis calls